src/backend/distributed/executor/multi_utility.c

```diff
@@ -1045,88 +1055,175 @@ ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString)
 
 /*
  * ExecuteCommandOnWorkerShards executes a given command on all the finalized
- * shard placements of the given table. If the remote command errors out on the
- * first attempted placement, the function returns false. Otherwise, it returns
- * true.
- *
- * If the remote query errors out on the first attempted placement, it is very
- * likely that the command is going to fail on other placements too. This is
- * because most errors here will be PostgreSQL errors. Hence, the function fails
- * fast to avoid marking a high number of placements as failed. If the command
- * succeeds on at least one placement before failing on others, then the list of
- * failed placements is returned in failedPlacementList.
+ * shard placements of the given table with two-phase commit.
+ * 
+ * ExecuteCommandOnWorkerShards opens an individual connection for each of the   
+ * shard placement. After all connections are opened, a BEGIN command followed by 
+ * a proper "SELECT worker_apply_shard_ddl_command(<shardId>, <DDL Command>)" is
+ * sent to all open connections in a serial manner.
+ * 
+ * The opened transactions are handled by the CompleteShardPlacementTransactions
+ * function.
  *
  * Note: There are certain errors which would occur on few nodes and not on the
  * others. For example, adding a column with a type which exists on some nodes
- * and not on the others. In that case, this function might still end up returning
- * a large number of placements as failed.
+ * and not on the others. 
+ * 
+ * Note: The execution will be blocked if a prepared transaction from previous
+ * executions exist on the workers. In this case, those prepared transactions should
+ * be removed by either COMMIT PREPARED or ROLLBACK PREPARED.
  */
 static bool
 ExecuteCommandOnWorkerShards(Oid relationId, const char *commandString,
 							 List **failedPlacementList)
 {
-	bool isFirstPlacement = true;
-	ListCell *shardCell = NULL;
-	List *shardList = NIL;
+	List *shardIdList = LoadShardList(relationId);
 	char *relationOwner = TableOwner(relationId);
-
-	shardList = LoadShardList(relationId);
-	foreach(shardCell, shardList)
+	MemoryContext oldContext = MemoryContextSwitchTo(TopTransactionContext);
+	HTAB *shardConnectionHash = OpenTransactionsToAllShardPlacements(shardIdList,
+																	relationOwner);
+	ListCell *shardCell = NULL;
+	
+	shardPlacementConnectionList = ConnectionList(shardConnectionHash);
+	MemoryContextSwitchTo(oldContext);
+	
+	if (!isXactCallbackRegistered)
+	{
+		RegisterXactCallback(CompleteShardPlacementTransactions, NULL);
+		isXactCallbackRegistered = true;
+	}
+	
+	foreach(shardCell, shardIdList)
```

Comment:
I am not sure but, I think using PQsendQuery instead of PQexec in ExecuteCommandOnShardPlacements function,  and checking the results after the foreach with `PQgetResult` would be needed.